packages/react-reconciler/src/ReactStrictModeWarnings.js

```diff
@@ -324,27 +324,49 @@ if (__DEV__) {
   ReactStrictModeWarnings.flushLegacyContextWarning = () => {
     ((pendingLegacyContextWarning: any): FiberToFiberComponentsMap).forEach(
       (fiberArray: FiberArray, strictRoot) => {
-        const uniqueNames = new Set();
+        const componentStacks = new Map();
+
         fiberArray.forEach(fiber => {
-          uniqueNames.add(getComponentName(fiber.type) || 'Component');
+          const componentName = getComponentName(fiber.type) || 'Component';
+          const componentStack = getStackByFiberInDevAndProd(fiber);
+          let count = 0;
+          if (componentStacks.has(componentStack)) {
+            ({count} = (componentStacks.get(componentStack): any));
+          }
+          // Increase count by 1
+          componentStacks.set(componentStack, {
+            count: count + 1,
+            name: componentName,
+            stack: componentStack,
+          });
           didWarnAboutLegacyContext.add(fiber.type);
         });
 
-        const sortedNames = setToSortedString(uniqueNames);
-        const strictRootComponentStack = getStackByFiberInDevAndProd(
-          strictRoot,
-        );
-
-        console.error(
-          'Legacy context API has been detected within a strict-mode tree.' +
-            '\n\nThe old API will be supported in all 16.x releases, but applications ' +
-            'using it should migrate to the new version.' +
-            '\n\nPlease update the following components: %s' +
-            '\n\nLearn more about this warning here: https://fb.me/react-legacy-context' +
-            '%s',
-          sortedNames,
-          strictRootComponentStack,
-        );
+        // Get the stacks from our componentStacks Map
+        const stacks = Array.from(componentStacks.values());
+
+        // Sort the stacks by their counts
+        stacks.sort((a, b) => b.count - a.count);
+
+        const mostFrequentStack = stacks[0];
+
+        if (mostFrequentStack) {
+          // We map to a Set to remove duplicate component names
+          const componentNames = Array.from(
+            new Set(stacks.map(stack => stack.name)),
+          ).join(', ');
```

Comment:
It's not a huge deal, since this is DEV-only code, but I wonder if there isn't a slightly more efficient way of doing this. Rather than creating all of the component stacks and then doing a full sort based on count, maybe we could just change `didWarnAboutLegacyContext` from a `Set` to a `Map` that counts the number of times each fiber would have warned, and then look for the fiber with the highest count and generate its stack only?

Reply:
I mentioned this to @gaearon and he mentioned that this wouldn't be reliable, as we should be using the stack instead because different fibers might make the same stack. In terms of effeciency this seemed fine given it was DEV, it's not ideal, but given we don't want to use the Fiber as the key, it seems sensible.

Reply:
> different fibers might make the same stack

Hmm...given that the stack is essentially just the name (and potentially source location) of each fiber from the current one through its parent path, the chance of two fibers generating the same stack seems unlikely? Am I misunderstanding something? Can you mention an example of where this might happen?

Reply:
I can't remember that exact details of what @gaearon said, he might be able to give a detailed explanation as it's late here and I'm a bit drained. However, if you had an alternate fiber, you'd run into this issue on updates (because we switch fiber) which would generate the same stack but would obviously be a referentially different fiber.

Reply:
Our check could be for `fiber` or `fiber.alternate` to account for this. (We do similar things in DevTools to account for this.)

I guess I'm curious to see what Dan says about this then.

Reply:
My thinking was that if we have 100 fibers with the same stack (e.g. list items), I'd like to their stacks to have 100 weight. Instead of each of them having 1 weight. Because fixing legacy context would have effect on 100 subtrees.

>maybe we could just change didWarnAboutLegacyContext from a Set to a Map that counts the number of times each fiber would have warned

Wouldn't each fiber only ever warn once? I'm not quite following this. Or do you mean a map by type?

I think a map from fiber type to count makes sense to me. "Fix the most impactful type first." You wouldn't see the most common *stack* for that type though, but maybe it's OK?

Reply:
Fair enough, I guess e.g. list items would have the same stack if we didn't have file name and line number info.

> Wouldn't each fiber only ever warn once? I'm not quite following this. Or do you mean a map by type?

Sorry, yes. I meant mapping by type.

Reply:
Sounds good.